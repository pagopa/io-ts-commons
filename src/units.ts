/**
 * Declares common unit types
 */

export interface IUnitTag<T> {
  readonly kind: T;
}

export type Millisecond = number & IUnitTag<"Millisecond">;

export type Second = number & IUnitTag<"Second">;
