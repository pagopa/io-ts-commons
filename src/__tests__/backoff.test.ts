import {
  calculateExponentialBackoffInterval,
  DEFAULT_EXPONENTIAL_BACKOFF_BASE_INTERVAL,
  DEFAULT_EXPONENTIAL_BACKOFF_MAX,
  DEFAULT_EXPONENTIAL_BACKOFF_MAX_JITTER,
  DEFAULT_EXPONENTIAL_BACKOFF_MULTIPLIER
} from "../backoff";
import { Millisecond } from "../units";

describe("exponentialBackoff", () => {
  it("should produce an exponentially increasing delay time", () => {
    const backoff = calculateExponentialBackoffInterval(
      DEFAULT_EXPONENTIAL_BACKOFF_BASE_INTERVAL,
      DEFAULT_EXPONENTIAL_BACKOFF_MULTIPLIER,
      0 as Millisecond,
      DEFAULT_EXPONENTIAL_BACKOFF_MAX
    );
    const tries = Array.from(Array(8).keys());
    const delays = tries.map(backoff);

    expect(delays).toMatchSnapshot();
  });

  it("should produce an exponentially increasing delay time with jitter", () => {
    const backoffNoJitter = calculateExponentialBackoffInterval(
      DEFAULT_EXPONENTIAL_BACKOFF_BASE_INTERVAL,
      DEFAULT_EXPONENTIAL_BACKOFF_MULTIPLIER,
      0 as Millisecond,
      DEFAULT_EXPONENTIAL_BACKOFF_MAX
    );
    const jitter = 1000 as Millisecond;
    const backoffWithJitter = calculateExponentialBackoffInterval(
      DEFAULT_EXPONENTIAL_BACKOFF_BASE_INTERVAL,
      DEFAULT_EXPONENTIAL_BACKOFF_MULTIPLIER,
      jitter,
      DEFAULT_EXPONENTIAL_BACKOFF_MAX
    );
    const tries = Array.from(Array(8).keys());
    const delaysNoJitter = tries.map(backoffNoJitter);
    const delaysWithJitter = tries.map(backoffWithJitter);

    // tslint:disable-next-line:no-let
    for (let i = 0; i < tries.length; i++) {
      const diff = delaysWithJitter[i] - delaysNoJitter[i];
      expect(diff).toBeGreaterThanOrEqual(0);
      expect(diff).toBeLessThanOrEqual(jitter);
    }
  });
});
