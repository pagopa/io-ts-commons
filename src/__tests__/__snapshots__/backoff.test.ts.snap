// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exponentialBackoff should produce an exponentially increasing delay time 1`] = `
Array [
  500,
  1000,
  2000,
  4000,
  8000,
  16000,
  32000,
  32000,
]
`;
